# Multiple Regression with `statsmodels`

A multiple linear regression is simply a linear regression that involves more than one predictor variable. It is represented as:
$$\qquad Y_e = \alpha + \beta_1*X_1  + \beta_2*X_2 + \dots  + \beta_p*X_p$$  

Each *β<sub>i</sub>* will be estimated using the least sum of squares method.

The data set is  
$$ \begin{array} 
       {~~}  Y_1, &  X_1^{(1)},  &  \ldots, &  X_p^{(1)} \\
        Y_2, &  X_1^{(2)},  &  \ldots, &  X_p^{(2)} \\
       \vdots  & \vdots  & \vdots & \vdots \\
      Y_n, &  X_1^{(n)},  &  \ldots, &  X_p^{(n)} 
    \end{array}
$$
For each sample $i$, the predicted value by the model is:  $\qquad Y_{i,e} = \alpha + \beta_1*X_1^{(i)}  + \beta_2*X_2^{(i)} + \dots  + \beta_p*X_p^{(i)}$  


Define the sum of squares 
$$   S(\alpha,\beta_1,\ldots,\beta_p) = \sum_{i=1}^n 
\left\{     Y_i -Y_{i,e}\right\}^2  =\sum_{i=1}^n \left\{ 
    Y_i -\left( \alpha + \beta_1*X_1^{(i)}  + \beta_2*X_2^{(i)} + \dots  + \beta_p*X_p^{(i)}\right)\right\}^2
$$
Least squares estimators: solve 
$$ \frac{\partial  S(\alpha,\beta_1,\ldots,\beta_p)}{\partial \alpha}=0,\quad 
\frac{\partial S (\alpha,\beta_1,\ldots,\beta_p)}{\partial \beta_1}=0,\quad \ldots,\quad
\frac{\partial S (\alpha,\beta_1,\ldots,\beta_p)}{\partial \beta_p}=0. 
$$
to obtain the `least squares estimators` of the parameters
$$ \hat\alpha, \hat\beta_1,\ldots,\hat\beta_p.
$$
Note that be definition, 
$$      SSD = S(\hat\alpha, \hat\beta_1,\ldots,\hat\beta_p).
$$
In other words, the fitted SSD (difference of sum of squares) is the minimized value of the sum squares with the estimated values of the parameters.


`The more varibles, the smaller the $R^2$`

Consider two regression models

(I)   $\quad ~ Y_e = \alpha + \beta_1*X_1$ 

(II)  $\quad  \tilde Y_e = \alpha + \beta_1*X_1  + \beta_2*X_2 $

The model (II) has one more input variable $X_2$. 

The $SSD_I$ of Model (I) is the minimum of

$$   S_I(\alpha,\beta_1) = \sum_{i=1}^n \left\{ 
    Y_i -\left( \alpha + \beta_1*X_1^{(i)} \right)\right\}^2
$$
over all possible values of $(\alpha,\beta_1)$.

The $SSD_{II}$ of Model (II) is the minimum of 

$$   S_{II}(\alpha,\beta_1,\beta_2) = \sum_{i=1}^n \left\{ 
    Y_i -\left( \alpha + \beta_1*X_1^{(i)} +\beta_2*X_2^{(i)}  \right)\right\}^2. 
$$
over all possible values of $(\alpha,\beta_1,\beta_2)$.

Because  $\quad S_I(\alpha,\beta_1) = S_{II}(\alpha,\beta_1,\beta_2=0 )$, 

we find that $SSD_{II}\le SSD_I$, so 
$$   R^2_{II} = SST - SSD_{II} \ge SST - SSD_{I} =  R^2_{I}.
$$

With this simple dataset of three predictor variables, there can be seven possible models:

1. Sales ~ TV
2. Sales ~ Newspaper
3. Sales ~ Radio
4. Sales ~ TV + Radio
5. Sales ~ TV + Newspaper
6. Sales ~ Newspaper + Radio
7. Sales ~ TV + Radio + Newspaper

Generally, if there are p possible predictor variables, there can be *(2<sup>p</sup> - 1)* possible models – this can get large very quickly!

Thankfully, there are a few guidelines to filter some of these and then navigate towards the most efficient one.

- Keep variables with low p-values and eliminate ones with high p-values
- Keep variables that increase the value of **adjusted-*R<sup>2</sup>*** – this penalizes the model for adding insignificant variables and increases when we add significant variables. It is calculated by: 
$$ R^2_{adj} = 1- (1-R^2) \frac{n-1}{n-p-1}$$

Based on these guidelines, there are two approaches to select the predictor variables in the final model:

- **Forward selection**: start with a null model (no predictors), then add predictors one by one. If the p-value for the variable is small enough and the value of the adjusted-*R<sup>2</sup>* goes up, the predictor is included in the model. Otherwise, it is not included.
- **Backward selection**: starts with a model that has all the possible predictors and discard some of them. If the p-value of a predictor variable is large and adjusted-*R<sup>2</sup>* is lower when removed, it is discarded from the model. Otherwise, it remains a part of the model.

Many statistical programs give us an option to select from these approaches while implementing multiple linear regression.

For now, let’s manually add a few variables and see how it changes the model parameters and efficacy. First, add the `newspaper` variable to the model:

In [1]:
# Import necessary libaries
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# Import and display first five rows of advertising dataset

github_path = 'https://raw.githubusercontent.com/huangyh09/foundation-data-science/'
dat_dir = github_path + 'main/w9-regression/'
# dat_dir = './'

advert = pd.read_csv('advertising.csv')
advert.head()

TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [2]:
# Initialise and fit new model with TV and Newspaper as predictors
model2 = smf.ols('Sales ~ TV + Newspaper', data=advert).fit()
print(model2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.646
Model:                            OLS   Adj. R-squared:                  0.642
Method:                 Least Squares   F-statistic:                     179.6
Date:                Mon, 01 Nov 2021   Prob (F-statistic):           3.95e-45
Time:                        13:41:13   Log-Likelihood:                -509.89
No. Observations:                 200   AIC:                             1026.
Df Residuals:                     197   BIC:                             1036.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      5.7749      0.525     10.993      0.0

As you see, the p-values for the coefficients are very small, suggesting that all the estimates are significant. The equation for this model will be:

 $$ \text{Sales} = 5.77+0.046* \text{TV} + 0.044 * \text{Newspaper}$$

The values of *R<sup>2</sup>* and adjusted *R<sup>2</sup>* are 0.646 and 0.642, which is just a minor improvement from before  (0.0612 and 0.0610, respectively).

To calculate the RSE:

In [3]:
# Store parameters
alpha = model2.params[0]
beta1 = model2.params[1]
beta2 = model2.params[2]

# Calculate RSE
advert['sales_pred'] = model2.predict()
advert['SSD'] = (advert['Sales'] - advert['sales_pred'])**2
SSD = advert['SSD'].sum()
RSE = np.sqrt(SSD / 197)   # n = 200, p = 2
salesmean = np.mean(advert['Sales'])
error = RSE / salesmean

print(f'RSE = {RSE}\nMean sale = {salesmean}\nError = {np.round(error, 4)*100}%')

RSE = 3.1207198602528856
Mean sale = 14.022500000000003
Error = 22.259999999999998%


Only a small decrease in RSE and error...

Let’s take a closer look at the summary above. The Adj-R<sup>2</sup> increases slightly, but the F-statistic decreases (from 312.1 to 179.6), as does the associated p-value. This suggests that adding `newspaper` didn't improve the model significantly.

Let's try adding `radio` instead:

In [4]:
# Initialise and fit new model with TV and Radio as predictors
model3 = smf.ols('Sales ~ TV + Radio', data=advert).fit()
print(model3.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Mon, 01 Nov 2021   Prob (F-statistic):           4.83e-98
Time:                        13:41:13   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9211      0.294      9.919      0.0

This gives us the model:

![](https://latex.codecogs.com/gif.latex?%5Ctext%7BSales%7D%3D2.92+0.046*%5Ctext%7BTV%7D+0.188*%5Ctext%7BRadio%7D)

The adjusted *R<sup>2</sup>* value has improved considerably, as did the F-statistic, indicating an efficient model.

The RSE can be calculated using the same method above:

In [5]:
# Store parameters
alpha = model3.params[0]
beta1 = model3.params[1]
beta2 = model3.params[2]

# Calculate RSE
advert['sales_pred'] = model3.predict()
advert['SSD'] = (advert['Sales'] - advert['sales_pred'])**2
SSD = advert['SSD'].sum()
RSE = np.sqrt(SSD / 197)   # n = 200, p = 2
salesmean = np.mean(advert['Sales'])
error = RSE / salesmean

print(f'RSE = {RSE}\nMean sale = {salesmean}\nError = {np.round(error, 4)*100}%')

RSE = 1.681360912508001
Mean sale = 14.022500000000003
Error = 11.99%


An improvement!

Thus, we can conclude that `radio` is a great addition to the model.  `TV` and `radio` advertising costs together are able to predict sales well. But, can we improve it a bit further by combining all three predictor variables?

See if you can figure out how to do this on your own!

In [6]:
# Initialise and fit new model with TV, Newspaper, and Radio as predictors

# Print summary of regression results

# Store parameters - there are now three betas

# Calculate RSE - don't forget that the number of predictors p is now 3


You should get the equation:

$$ \text{Sales} = 2.939+0.046*\text{TV} -0.001*\text{Newspaper} +0.188*\text{Radio}$$


You should also find that:

- RSE increases slightly,
- the coefficient for `newspaper` is negative, and
- the F-statistic decreases considerably from 859.6 to 570.3.

All these suggest that the model actually became less efficient on addition of `newspaper`. 

Why?

This step shows clearly that adding one more input variable `Newspaper` in Model 3 does not lead to any improvement. 

Here we manually checked a few combanations of input variables to reach an acceptable choice. There are more systematic approaches to select the predictor variables in a final model:

- **Forward selection**: start with a null model (no predictors), then add predictors one by one. If the p-value for the variable is small enough and the value of the adjusted-*R<sup>2</sup>* goes up, the predictor is included in the model. Otherwise, it is not included.
- **Backward selection**: starts with a model that has all the possible predictors and discard some of them. If the p-value of a predictor variable is large and adjusted-*R<sup>2</sup>* is lower when removed, it is discarded from the model. Otherwise, it remains a part of the model.

Many statistical programs give us an option to select from these approaches while implementing multiple linear regression.